packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,243 @@
+        this.libp2pComponents.connectionManager
+          .getConnections(peerId)
+          .forEach(
+            (conn) => (conn.tags = Array.from(new Set([...conn.tags, ...tags])))
+          );
+
+        this.dialAttemptsForPeer.delete(peerId.toString());
+        return;
+      } catch (error) {
+        log(`
+          Error dialing peer ${peerId.toString()}`);
+        dialAttempt = this.dialAttemptsForPeer.get(peerId.toString()) ?? 1;
+        this.dialAttemptsForPeer.set(peerId.toString(), dialAttempt + 1);
+
+        if (dialAttempt <= maxDialAttemptsForPeer) {
+          log(`Reattempting dial (${dialAttempt})`);
+        }
+      }
+    }
+
+    try {
+      log(`Deleting undialable peer ${peerId.toString()} from peer store`);
+      return await this.libp2pComponents.peerStore.delete(peerId);
+    } catch (error) {
+      throw `Error deleting undialable peer ${peerId.toString()} from peer store - ${error}`;
+    }
+  }
+
+  private startPeerDiscoveryListener(): void {
+    this.libp2pComponents.peerStore.addEventListener(
+      "peer",
+      this.onEventHandlers["peer:discovery"]
+    );
+  }
+
+  private startPeerConnectionListener(): void {
+    this.libp2pComponents.connectionManager.addEventListener(
+      "peer:connect",
+      this.onEventHandlers["peer:connect"]
+    );
+  }
+
+  private startPeerDisconnectionListener(): void {
+    // TODO: ensure that these following issues are updated and confirmed
+    /**
+     * NOTE: Event is not being emitted on closing nor losing a connection.
+     * @see https://github.com/libp2p/js-libp2p/issues/939
+     * @see https://github.com/status-im/js-waku/issues/252
+     *
+     * >This event will be triggered anytime we are disconnected from another peer,
+     * >regardless of the circumstances of that disconnection.
+     * >If we happen to have multiple connections to a peer,
+     * >this event will **only** be triggered when the last connection is closed.
+     * @see https://github.com/libp2p/js-libp2p/blob/bad9e8c0ff58d60a78314077720c82ae331cc55b/doc/API.md?plain=1#L2100
+     */
+    this.libp2pComponents.connectionManager.addEventListener(
+      "peer:disconnect",
+      this.onEventHandlers["peer:disconnect"]
+    );
+  }
+
+  private onEventHandlers = {
+    "peer:discovery": async (evt: CustomEvent<PeerInfo>): Promise<void> => {
+      const { id: peerId } = evt.detail;
+      if (!(await this.shouldDialPeer(peerId))) return;
+
+      this.dialPeer(peerId).catch((err) =>
+        log(`Error dialing peer ${peerId.toString()} : ${err}`)
+      );
+    },
+    "peer:connect": (evt: CustomEvent<Connection>): void => {
+      {
+        this.keepAliveManager.start(
+          evt.detail.remotePeer,
+          this.libp2pComponents.ping.bind(this)
+        );
+      }
+    },
+    "peer:disconnect": () => {
+      return (evt: CustomEvent<Connection>): void => {
+        this.keepAliveManager.stop(evt.detail.remotePeer);
+      };
+    },
```

Comment:
I am not sure why the connection manager is encumbered with keep alive manager logic.
Can't the keep alive manager watch events it needs to react too itself?